### ISSUE TITLE
Right CTRL + ALT + any key combination does not work 

### ISSUE THREAD

**Issue by victorhge** [mintty/mintty#969]
mintty 3.0.5 3.0.6, 3.1.0
I run emacs inside mintty in text mode.  
Left ctrl+ left alt +any key works.  But Right Ctrl+ Right Alt +any key combination does not get any response. 

There is no such problem on mintty 2.9.6.

**Comment by mintty** [comment]
Please be more specific. Which key combination should give you which response? Did you configure anything specific in emacs about this?
Also, what is your keyboard layout, and which is your respective cygwin version? If you downgrade cygwin to 3.0.7 with current mintty, does the problem persist (asking because of specific changes in that transition)?

**Comment by mintty** [comment]
A reproducible test case is needed, or I can only close the issue.

**Comment by mintty** [comment]
Also, is the issue (whichever...) reproducible with Right-Ctrl and Left-Alt?

**Comment by mintty** [comment]
Also, is the issue (whichever...) reproducible with Left-Ctrl and Right-Ctrl and Right-Alt?
(Asking to check whether it's reproducible on a non-US keyboard.)

**Comment by mintty** [comment]
If an issue at all, it may have slipped in with 3.0.1. In any case, a cooperating tester is needed to narrow it down and further debug it. Otherwise closing soon.

**Comment by victorhge** [comment]
> 
> 
> Please be more specific. Which key combination should give you which response? Did you configure anything specific in emacs about this?
For instance,  if you press C-h k, then press anything, it will display the document of the function invoked by the key combination.   But there is no reponse at all for right ctrl + right +alt +any key.

Emacs is run as "emacs -q -nw" which means nothing is specific.
> Also, what is your keyboard layout, and which is your respective cygwin version? If you downgrade cygwin to 3.0.7 with current mintty, does the problem persist (asking because of specific changes in that transition)?
$ uname -a
CYGWIN 3.1.4(0.340/5/3) 2020-02-19 08:49 x86_64 Cygwin. 

I am not going to downgrade Cygwin.

**Comment by victorhge** [comment]
> 
> 
> A reproducible test case is needed, or I can only close the issue.

What does a test case look like? Would you give an example.  Can you reproduce this problem?

**Comment by victorhge** [comment]
> 
> 
> Also, is the issue (whichever...) reproducible with Right-Ctrl and Left-Alt?

Never tried.

**Comment by victorhge** [comment]
> 
> 
> Also, is the issue (whichever...) reproducible with Right-Ctrl and Left-Alt?

Never tried.

**Comment by mintty** [comment]
> > A reproducible test case is needed, or I can only close the issue.

> What does a test case look like? Would you give an example. Can you reproduce this problem?

I would not have asked so many questions if I were able to reproduce an issue.
In the first place, I would expect you to describe what actually happens. Pick one specific key combination (e.g. Right-Ctrl+Right-Alt+x), describe what you expect and what you see.
Repeat on the command line, after Ctrl+V (to quote control characters literally).
Repeat with the other Ctrl/Alt combinations I asked about.
Only then we could talk about an "issue" at all.

In a next step, I may ask you to try with mintty 3.0.0 vs. 3.0.1 or other versions, depending on previous results. If that doesn't get us further, I may have to ask you to downgrade cygwin for a test case. No problem to upgrade back afterwards, but that's life. If we can't identify an issue, there is none.

**Comment by scandum** [comment]
It's easy to reproduce with TinTin++
```
tt++
#config  convert on
```
Right-Ctrl + Right-Alt combinations give no input. Individually the two keys work as expected.

It seems to be caused by \e[>4;1m
```
tt++
#showme \e[>4n
#config convert on
```
Everything works as expected.

**Comment by mintty** [comment]
So maybe that application, or its server, has changed to send that sequence?
If there is claim something changed in mintty, I'd still need response to my diagnostic questions.
What exactly do you type in (rather than fuzzy "combinations")?
And check whether something different happens now in mintty 2.9.6.

**Comment by mintty** [comment]
OK, in modifyOtherKeys mode 1, Ctrl+Right-Alt+letter used to enter ESC Ctrl+letter until mintty 2.9.6 and no more with 2.9.7. On the other hand, Ctrl+Shift+letter did not work before and was fixed in 2.9.7.

**Comment by scandum** [comment]
Walkthrough for bash on Cygwin:
```
echo -e "\e[>4;1m"
ctrl-v
right ctrl + alt + x
(nothing happens)
echo -e "\e[>4n"
ctrl-v
right ctrl + alt + x
(^[ is printed)
```

**Comment by mintty** [comment]
> (^[ is printed)

Actually, ^[^X is sent, and then interpreted by the shell.
Thanks.

**Comment by victorhge** [comment]
> 
> 
> Walkthrough using bash on Cygwin (mintty 3.1.0):
> 
> ```
> echo -e "\e[>4;1m"
> ctrl-v
> right ctrl + alt + x
> (nothing happens)
> echo -e "\e[>4n"
> ctrl-v
> right ctrl + alt + x
> (^[ is printed)
> ```

emacs does send "\e[>4;1m" to mintty to enable modifyOtherKeys  mode.

**Comment by mintty** [comment]
Could reproduce the issue on one machine, but not on another. Puzzled.
Therefore one additional diagnostic question: if your keyboard has a right Alt key, does it make a difference to use
right ctrl + left alt + x
vs
right ctrl + right als +x
?

**Comment by scandum** [comment]
In my case the problem is definitely with the right alt. The left alt always works.

right ctrl + right alt + x - doesn't work
left ctrl + right alt + x - doesn't work

**Comment by mintty** [comment]
Note the AltGr key on international keyboards. As far as I know, the right Alt key on US keyboards has the same function and is actually incorrectly labelled identical to the left Alt key.
So basically, AltGr+x should in fact *not* produce any input, so the behaviour you describe is correct.
The confusing and misleading thing is that mintty without modifyOtherKeys mode currently *does* provide input even for combinations that do not have a mapping for AltGr as defined by the effective keyboard layout, e.g. AltGr+x inputs ESC x. (Compare xterm which does not do that.)
This is inconsistent and I've commited a change to stop this undesirable behaviour.
I am aware it may not be the change you had hoped for, but now mintty is consistent and compatible about AltGr handling.

**Comment by scandum** [comment]
I personally don't care, but I'd like to add that if anyone needs this behavior they should take it up with the xterm developer.

**Comment by BrianInglis** [comment]
Right Alt on non-International English keyboard layouts works just like Left Alt, so no accents, dead keys, or compose function is available, other than Alt numpad and other OEM and Unicode numeric character input methods, unless those keys are intercepted and interpreted by the app in focus.

**Comment by mintty** [comment]
@scandum reported above that right alt and left alt behave differently. What keyboard and keyboard layout was that tested with?
Can you built a debug version of mintty yourself? In file wininput.c please active `#define debug_virtual_key_codes` (remove the `dont_` prefix). Run that mintty from a command line (other mintty). What output do left-alt vs. right-alt produce with a few combinations like 'x', Ctrl+'x'?

**Comment by feryardiant** [comment]
I had similar issue https://github.com/mskyaxl/wsl-terminal/issues/148 

As mentioned, since wsl-terminal `v0.8.12` (which is using mintty `v2.9.5`) I no longer can go fullscreen using <kbd>RIGHT-ALT</kbd> + <kbd>ENTER</kbd> like wsl-terminal `v0.8.10` (mintty `v2.8.4`) was.

**Comment by mintty** [comment]
Right-Alt+Enter is not supposed to switch full-screen mode.

**Comment by feryardiant** [comment]
So, is it possible to other package that use mintty (e.g. wsl-terminal, etc) to bind <kbd>RIGHT-ALT</kbd> / <kbd>RIGHT-CTRL</kbd> + <kbd>any</kbd> as desired? So, both <kbd>RIGHT-ALT</kbd> / <kbd>RIGHT-CTRL</kbd> could behave the same as <kbd>LEFT-ALT</kbd> / <kbd>LEFT-CTRL</kbd>?

**Comment by mintty** [comment]
You can define functions or text insertions for combinations like Alt+Enter, Alt+Ctrl+Enter with the KeyFunctions option, however, right-Alt does not work because it has a special function for the AltGr key which extends the layout of international keyboards.

**Comment by feryardiant** [comment]
Thank you @mintty for explanation.

**Comment by jlennox** [comment]
> Right-Alt+Enter is not supposed to switch full-screen mode.

This is confusing UI. I spent nearly a year just assuming this key combination was broken because my actions were following what was written. It being described in the UI as "LeftAlt+Enter" or supporting right alt (I'm not sure why it isn't?) would remediate that.

**Comment by mintty** [comment]
> This is confusing UI.

Yes, but it's the Windows UI. On international keyboards, there is no right Alt key. Instead there is an AltGr key which is a modifier to generate special characters as used in the respective keyboard layout. In their peculiar wisdom, MS decided that AltGr should have the same keycode as right Alt (in some tricky combination with left Ctrl), that's where the trouble comes from.

Agreed, I could adapt the manual to make that clear. I could also support right-Alt+Enter as a special case, but then another person would claim right-Alt+letter should work too, see the original problem of this issue.

**Comment by mintty** [comment]
Added an explanation to the manual.

**Comment by feryardiant** [comment]
> I could also support right-Alt+Enter as a special case, but then another person would claim right-Alt+letter should work too

@mintty I can understand the situation but I do believe there must be an Windows API that let another program know whether or not the _international keyboard_ enabled, so the program can behave exactly what users expected.

Anyway, AFAIK, in [Windows Terminal](https://github.com/Microsoft/Terminal) any combination of Left <kbd>CTRL</kbd> or <kbd>ALT</kbd> + <kbd>Key</kbd> works the same as Right <kbd>CTRL</kbd> or <kbd>ALT</kbd>. At least in my use-cases. Also, I found some people who expects <kbd>CTRL+ALT+Key</kbd> should behave <kbd>AltGr+Key</kbd> here https://github.com/microsoft/terminal/issues/5525 and either Left or Right <kbd>CTRL</kbd> or <kbd>ALT</kbd> should behave the same as well here https://bugzilla.mozilla.org/show_bug.cgi?id=1389739. Lastly, I can't found any combinations on any keyboard layouts of <kbd>AltGr+Enter</kbd>, correct me if I'm wrong.

That said, that any combination of Left or Right <kbd>CTRL+ALT+Key</kbd> is <kbd>AltGr+Key</kbd> and if there's no <kbd>AltGr+Key</kbd> combination it shouldn't be treated as an input. Is that make sense?

Thoughts?

**Comment by mintty** [comment]
>  CTRL+ALT+Key should behave AltGr+Key

Mintty has an option "Ctrl+LeftAlt is AltGr".
However, such a setting spoils the usage of Ctrl and Alt as independent modifiers that can also be combined, which is desirable to implement flexible keyboard functions.
Given that 1. AltGr must be supported and 2. Ctrl+Alt should be combinable and even 3. as AltGr combinations are normal characters (some even ASCII) also Alt+AltGr should be a useful combination (e.g. to generate an ESC prefix for the @ character), it is not possible to handle left/right the same.
Your final proposition would simplify the capabilities of keyboard layout too much.

**Comment by mintty** [comment]
> microsoft/terminal#5525

Commented in that issue that I consider that a wrong approach.

**Comment by BrianInglis** [comment]
Somehow MS/Windows often manages to choose suboptimal designs focussing, on adding new features (great new feature/UX) without apparent consideration of how to do so with minimal changes to previous/current features/UX. 
A better approach than the Windows International layouts, which change the keyboard layouts to default a lot of punctuation keys to dead keys, which may be okay if users rarely need to use the punctuation, but requires you to type **`SPACE`** to get the base punctuation characters, and makes writing code for almost every language slower and more painful, appears to be *Scott Severance*'s
[Windows version of the Ubuntu “English (intl., with AltGr dead keys)” keyboard layout](https://www.scottseverance.us/html/keyboard/win_intl_altgr/)
which adds no dead keys but **`AltGr`** as a **`COMPOSE`** key you have to press to add accents.

Copies of the links in his article to:
- [View Source](https://www.scottseverance.us/html/keyboard/win_intl_altgr/intAltGr.klc.html)
- [Download Source](https://www.scottseverance.us/html/keyboard/win_intl_altgr/intAltGr.klc)
- [Download Installer and Support Files](https://www.scottseverance.us/html/keyboard/win_intl_altgr/intaltgr.zip)
- [Download Microsoft Keyboard Layout Creator original](https://www.microsoft.com/en-us/download/details.aspx?id=22339&ranMID=43674&ranEAID=je6NUbpObpQ&ranSiteID=je6NUbpObpQ-F._7NJb45556hA3Hd8I_zQ&epi=je6NUbpObpQ-F._7NJb45556hA3Hd8I_zQ&irgwc=1&OCID=AID681541_aff_7795_1243925&tduid=(ir__joa0lm6kekkfrwkkxkyc9pw9y32xhqfjx6stmyt900)(7795)(1243925)(je6NUbpObpQ-F._7NJb45556hA3Hd8I_zQ)()&irclickid=_joa0lm6kekkfrwkkxkyc9pw9y32xhqfjx6stmyt900) [updated Windows 10 release](https://www.microsoft.com/en-us/download/details.aspx?id=22339) to modify, develop, or install alternative layouts.
Some of the files are in *UTF-16LE* and must remain in that encoding to work with the MS tools: only use `d2u` in *newfile* mode with an output file name: **`$`**`d2u -ul -n intAltGr.klc intAltGr.txt`

**Comment by mintty** [comment]
Released 3.1.5.

**Comment by mintty** [comment]
Released 3.1.7.
